File: benchmark/14_champernowne/champernowne.go
// Benchmark 14: Champernowne Constant
// Category: Number Theory / String Processing
// Tests: Number to string conversion, string concatenation, digit counting

package main

import (
	"fmt"
	"strconv"
	"time"
)

const MAX_N = 1000000 // Concatenate numbers 1 to 1,000,000

func main() {
	start := time.Now()

	// Build buffer using byte slice
	buffer := make([]byte, 0, 6000000)

	// Concatenate all numbers from 1 to MAX_N
	for i := 1; i <= MAX_N; i++ {
		buffer = strconv.AppendInt(buffer, int64(i), 10)
	}

	// Count each digit 0-9
	digitCounts := make([]int, 10)
	for _, b := range buffer {
		digitCounts[b-'0']++
	}

	elapsed := time.Since(start)

	fmt.Printf("Numbers concatenated: 1 to %d\n", MAX_N)
	fmt.Printf("Total length: %d characters\n", len(buffer))
	fmt.Printf("Digit counts:\n")
	for d := 0; d < 10; d++ {
		fmt.Printf("  %d: %d\n", d, digitCounts[d])
	}
	fmt.Printf("Time: %.4f seconds\n", elapsed.Seconds())
}

File: benchmark/14_champernowne/champernowne.rs
// Benchmark 14: Champernowne Constant
// Category: Number Theory / String Processing
// Tests: Number to string conversion, string concatenation, digit counting

use std::time::Instant;

const MAX_N: usize = 1_000_000; // Concatenate numbers 1 to 1,000,000

fn main() {
    let start = Instant::now();

    // Build buffer using String
    let mut buffer = String::with_capacity(6_000_000);

    // Concatenate all numbers from 1 to MAX_N
    for i in 1..=MAX_N {
        buffer.push_str(&i.to_string());
    }

    // Count each digit 0-9
    let mut digit_counts = [0usize; 10];
    for c in buffer.chars() {
        digit_counts[c.to_digit(10).unwrap() as usize] += 1;
    }

    let elapsed = start.elapsed();

    println!("Numbers concatenated: 1 to {}", MAX_N);
    println!("Total length: {} characters", buffer.len());
    println!("Digit counts:");
    for d in 0..10 {
        println!("  {}: {}", d, digit_counts[d]);
    }
    println!("Time: {:.4} seconds", elapsed.as_secs_f64());
}
